### PR TITLE
Add support of input object types

### LINKED ISSUE #31: [Code Generation] Handle InputObjectType objects from the typesUsed array
See #13 for more details

### PR DESCRIPTION
**Pull request by sav007**
Closes #31 

Refactoring query variables generation: remove map

Example of generate input type: 

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java

Example of usage:
```
ReviewInput.builder()
        .commentary("test")
        .stars(5)
        .favoriteColor(ColorInput.builder()
            .red(255)
            .green(255)
            .blue(255)
            .build()
        ).build();
```

@felipecsl @martijnwalraven @marwanad pls take a look

### CODE REVIEW

File: apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package com.example.input_object_type;
+
+import com.apollostack.api.Query;
+import java.lang.Override;
+import java.lang.String;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public final class TestQuery implements Query<TestQuery.Variables> {
```
Comment (marwanad) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java]: Should we be calling these mutations instead? We should use `operationType` field in the IR.

Not sure if it's worth renaming the Query interfaces to a `GraphQLOperation` interface. Then, having `GraphQLQuery` and `GraphQLMutation` extend those.
It's mainly for clarity since there's no real difference between queries and mutations other than the naming.

Comment (martijnwalraven) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java]: That's what we do on iOS as well, `GraphQLQuery` and `GraphQLMutation` extend `GraphQLOperation`.

Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java]: Do we need to distinguish between `GraphQLQuery` and `GraphQLMutation`. I checked iOS, `GraphQLQuery`, `GraphQLMutation` are just empty extension of `GraphQLOperation` without any additional functionality, so difference is only in naming for clarity, right?

Comment (martijnwalraven) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java]: On iOS, `ApolloClient` defines `fetch(query:)`, which takes a `GraphQLQuery`, and `perform(mutation:)`, which takes  a `GraphQLMutation`. There is no difference in implementation right now, but the idea is that we do need to differentiate between operation types because they support different options. Mutations shouldn't go through the cache for example, and it doesn't make sense to `watch` mutations.

Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java]: Oh I see, yes makes sense, thx will update

File: apollo-compiler/src/test/kotlin/com/apollostack/compiler/GraphQLCompilerTest.kt [hunk 2]
```diff
@@ -253,4 +253,24 @@ class GraphQLCompilerTest {
...
+
+    compiler.write(irFileFor("input_object_type", "TestQuery"))
+    assertThat(actual.readText()).isEqualTo(expected.readText())
+
+    val episode = JavaFileObjects.forSourceLines("com.example.input_object_type.Episode",
+        actualFileFor("input_object_type", "Episode").readLines())
+    val reviewInput = JavaFileObjects.forSourceLines("com.example.input_object_type.ReviewInput",
+        actualFileFor("input_object_type", "ReviewInput").readLines())
+    val colorInput = JavaFileObjects.forSourceLines("com.example.input_object_type.ColorInput",
+        actualFileFor("input_object_type", "ColorInput").readLines())
```
Comment (felipecsl) [inline apollo-compiler/src/test/kotlin/com/apollostack/compiler/GraphQLCompilerTest.kt]: Did you forget to add the expected files for `ColorInput`, `ReviewsInput`, `Episode` and assert on their contents?

Comment (sav007) [inline apollo-compiler/src/test/kotlin/com/apollostack/compiler/GraphQLCompilerTest.kt]: Added

**Comment by sav007** [comment]
@martijnwalraven @marwanad @felipecsl I update PR, added `GraphQLMutation`, `GraphQLQuery`, `GraphQLOperation`

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/InputObjectTypeSpecBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  companion object {
+    private fun TypeDeclarationField.graphQLType() = GraphQLType.resolveByName(type, !type.endsWith("!"))
+
+    private fun Number.castTo(graphQLType: GraphQLType) =
+      if (graphQLType is GraphQLType.GraphQLInt) {
+        toInt()
+      } else if (graphQLType is GraphQLType.GraphQLFloat) {
+        toFloat()
```
Comment (marwanad) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/InputObjectTypeSpecBuilder.kt]: should this be `toDouble`?

**Comment by sav007** [comment]
@felipecsl @martijnwalraven @marwanad if no objections I'm going to merge this PR

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      return this;
+    }
+
+    public Builder favoriteColor(@Nonnull ColorInput favoriteColor) {
+      reviewInput.favoriteColor = favoriteColor;
+      return this;
+    }
+
+    public ReviewInput build() {
+      return reviewInput;
```
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: I believe we should add validation to the `Builder#build()` methods to `throw IllegalStateException()` if any of the `@Nonnull` fields are `null`, right?

File: apollo-compiler/src/test/graphql/com/example/input_object_type/TestQuery.json [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      "fields": [
+        {
+          "name": "stars",
+          "description": "0-5 stars",
+          "type": "Int",
+          "defaultValue": null
+        },
+        {
+          "name": "commentary",
+          "description": "Comment about the movie, optional",
```
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQuery.json]: We should add the description as javadocs. Fine to do it in a separate PR

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.lang.Double;
+import javax.annotation.Nullable;
+
+public final class ColorInput {
+  int red = 1;
+
+  @Nullable Double green = 0.0;
+
+  double blue = 1.5;
```
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java]: all the fields should be final

Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java]: Instead of assigning the fields directly, the builder should initialize them in a constructor

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    private final ColorInput colorInput = new ColorInput();
+
+    Builder() {
```
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java]: private

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+package com.example.input_object_type;
+
+import java.lang.Integer;
+import java.lang.String;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public final class ReviewInput {
```
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: why make it final?

Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: Why does user need to extend it?

File: apollo-compiler/src/test/graphql/com/example/input_object_type/TestQuery.json (hunk 5 above)
Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/TestQuery.json]: #58 has been created

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java (hunk 7 above)
Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java]: well, we can but that not efficient as it will add additional (access$) method to class how is trying to get access to this constructor. It's recommended to have package protected scope for nested classes than private

https://realm.io/news/360andev-jake-wharton-java-hidden-costs-android/

Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java]: great point, 👍

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java (hunk 8 above)
Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: You might want to add more methods to it, like business logic related to your app.
I understand it might not be the best approach but I can see how people are gonna request this in the future for sure.
Fine to keep it this way for now, but it's gonna bite us in the future

**Comment by sav007** [comment]
@felipecsl I've updated PR:

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/ColorInputExpected.java

https://github.com/apollostack/apollo-android/blob/5661867bcf98c630403289c02b728c99887e27d8/apollo-compiler/src/test/graphql/com/example/input_object_type/TestQueryExpected.java

File: apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java (hunk 8 above)
Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: User should select composition over inheritance if he needs to add additional logic :)

Comment (felipecsl) [inline apollo-compiler/src/test/graphql/com/example/input_object_type/ReviewInputExpected.java]: I definitely agree
